Adding KeyHive Maypad

codybender opened the pull request:
Added the source files for the Maypad keyboard from KeyHive

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Added files under keyboards/KeyHive/Maypad

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

fauxpark commented on keyboards/KeyHive/Maypad/Maypad.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef KB_H
```
It appears you've generated these files with kbfirmware - their version of QMK is very old and basically not supported anymore. I would highly recommend you use the `./util/new_keyboard.sh` script to create the keyboard from the template already in this repo: https://docs.qmk.fm/#/hardware_avr?id=adding-your-avr-keyboard-to-qmk

codybender replied:
Awesome, thank you! Sorry, my first time here. I will generate it that way and make sure it all works! Thanks for the help

fauxpark replied:
No worries, it's just much easier on the reviewers (and you!) 😄

fauxpark commented on keyboards/KeyHive/Maypad/rules.mk:
```diff
@@ -0,0 +1,81 @@
+# MCU name
+#MCU = at90usb1286
```
```suggestion
```

fauxpark commented on keyboards/KeyHive/Maypad/Maypad.c:
```diff
@@ -0,0 +1,51 @@
+/* Copyright 2019 codybender
```
The keyboard and vendor directories should be all lowercase, to avoid problems with case-insensitive OSes.

fauxpark commented on keyboards/KeyHive/Maypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,74 @@
+
+// Defines the keycodes used by our macros in process_record_user
+enum custom_keycodes {
+  QMKBEST = SAFE_RANGE,
```
If you're not using these keycodes, this enum and `process_record_user()` can be removed.

fauxpark commented on keyboards/KeyHive/Maypad/Maypad.h:
```diff
@@ -0,0 +1,40 @@
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+#define LAYOUT( \
```
```suggestion
#define LAYOUT_ortho_5x4( \
```

It looks like this board also supports the `LAYOUT_numpad_5x4` layout.

fauxpark commented on keyboards/KeyHive/Maypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,74 @@
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	KEYMAP(
```
```suggestion
	LAYOUT_ortho_5x4(
```

fauxpark commented on keyboards/KeyHive/Maypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,74 @@
+		KC_P1, KC_P2, KC_P3, KC_ENT, 
+		KC_P0, KC_P0, KC_PDOT, KC_ENT),
+
+	KEYMAP(
```
```suggestion
	LAYOUT_ortho_5x4(
```

fauxpark commented on keyboards/KeyHive/Maypad/rules.mk:
```diff
@@ -0,0 +1,81 @@
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no 		# Enable support for HD44780 based LCDs (+400)
```
```suggestion
HD44780_ENABLE = no 		# Enable support for HD44780 based LCDs (+400)

LAYOUTS = ortho_5x4 numpad_5x4
```

fauxpark commented:
Also, I think the Honeycomb move might be better done in a separate PR, but I'm not sure. Let's see what the collaborators say about it.

codybender commented:
Awesome, thanks so much for your help @fauxpark. Sorry for the mishaps! Hopefully we are getting there. I fixed the filepaths hopefully here, and moved back the honeycomb.

fauxpark commented on keyboards/keyhive/maypad/keymaps/keymap.c:
```diff
@@ -0,0 +1,47 @@
+		KC_P1, KC_P2, KC_P3, KC_ENT, 
+		KC_P0, KC_P0, KC_PDOT, KC_ENT),
+
+	LAYOUT_ortho_5x4(
```
This layer actually doesn't need to exist as it's completely transparent and there's no way to get to it.

fauxpark commented on keyboards/keyhive/maypad/maypad.h:
```diff
@@ -0,0 +1,40 @@
+	{ K20,   K21,   K22,   K23 }, \
+	{ K30,   K31,   K32,   K33 }, \
+	{ K40,   K41,   K42,   K43 }  \
+}
```
You'll want to define `LAYOUT_numpad_5x4` here, and replace the missing spots with `KC_NO` in the second part of the macro. See the `clueboard/17` keyboard for an example.

drashna replied:
Also, it looks like make needs to have a newline at the end of the file here.  Without it, it just isn't compiling.  

```suggestion
}

```

And in general, we like to have a trailing line feed at the end of each file, probably for this exact reason.

fauxpark commented:
There are still some changes made to the Honeycomb (and the file modes are also still changed, though they should be 644 anyway).

And it would be a good idea to fill out the [info.json file](https://docs.qmk.fm/#/reference_info_json), for [Configurator](https://config.qmk.fm/) support. A decent method to do this is to recreate the layouts in [keyboard-layout-editor](http://www.keyboard-layout-editor.com/), paste the raw JSON into [qmk.fm/converter](https://qmk.fm/converter/), and paste *that* into info.json and fill out the relevant details. (It's a bit convoluted, hopefully there will be a more user friendly way at some point 😄)

noroadsleft commented:
As @fauxpark suggested, any changes to the Honeycomb should be in a separate pull request.

codybender commented:
Pulled latest commits from the master branch, and incorporated the changes above. seems to be working well now hopefully. sorry for the confusion and errors! Hopefully I fixed the deltas on the honeycomb by just pulling again.

noroadsleft commented on keyboards/keyhive/maypad/info.json:
```diff
@@ -0,0 +1,12 @@
+    "width": 4, 
+    "height": 5, 
+    "layouts": {
+        "LAYOUT": {
```
This key needs to be an exact match for the macro name from `maypad.h`.

```suggestion
        "LAYOUT_numpad_5x4": {
```

noroadsleft commented on keyboards/keyhive/maypad/info.json:
```diff
@@ -0,0 +1,12 @@
+    "height": 5, 
+    "layouts": {
+        "LAYOUT": {
+            "layout": [{"label":"Num Lock", "x":0, "y":0}, {"label":"/", "x":1, "y":0}, {"label":"*", "x":2, "y":0}, {"label":"-", "x":3, "y":0}, {"label":"7", "x":0, "y":1}, {"label":"8", "x":1, "y":1}, {"label":"9", "x":2, "y":1}, {"label":"+", "x":3, "y":1, "h":2}, {"label":"4", "x":0, "y":2}, {"label":"5", "x":1, "y":2}, {"label":"6", "x":2, "y":2}, {"label":"1", "x":0, "y":3}, {"label":"2",  ...
```
The JSON objects here need to be in the same order as the arguments used by `keymap.c` and `maypad.h` for this macro.

```suggestion
            "layout": [{"label":"Num Lock", "x":0, "y":0}, {"label":"/", "x":1, "y":0}, {"label":"*", "x":2, "y":0}, {"label":"-", "x":3, "y":0}, {"label":"7", "x":0, "y":1}, {"label":"8", "x":1, "y":1}, {"label":"9", "x":2, "y":1}, {"label":"4", "x":0, "y":2}, {"label":"5", "x":1, "y":2}, {"label":"6", "x":2, "y":2}, {"label":"+", "x":3, "y":1, "h":2}, {"label":"1", "x":0, "y":3}, {"label":"2", "x":1, "y":3}, {"label":"3", "x":2, "y":3}, {"label":"0", "x":0, "y":4, "w":2}, {"label":".", "x":2, "y":4}, {"label":"Enter", "x":3, "y":3, "h":2}]
```

noroadsleft commented on keyboards/keyhive/maypad/keymaps/keymap.c:
```diff
@@ -0,0 +1,47 @@
+/* Copyright 2019 codybender
```
This file needs to be in a subfolder of `keyhive/maypad/keymaps/` in order to work, and the layout macro `LAYOUT_ortho_5x4` doesn't currently exist in your codebase, so this file wouldn't compile in its current state.

codybender replied:
I believe I have fixed this in my first commits today

codybender replied:
Ah my apologies, I see what you meant now. I forgot I left those files hanging in /keymaps/. I have resolved this now

noroadsleft commented on keyboards/keyhive/maypad/keymaps/readme.md:
```diff
@@ -0,0 +1 @@
+# The default keymap for maypad
```
This file should be in the same subdirectory as the companion `keymap.c`.

codybender replied:
this as well is fixed in an earlier commit

noroadsleft commented on keyboards/keyhive/maypad/maypad.h:
```diff
@@ -0,0 +1,50 @@
+}
+
+#define LAYOUT LAYOUT_numpad_5x4
+#define KEYMAP LAYOUT_numpad_5x4
```
`KEYMAP` is a deprecated term for this usage. This line should be deleted.

```suggestion
```

noroadsleft commented:
@codybender I'm still seeing deltas on the Honeycomb files, and as you've used your master branch this might not be easily fixable.

QMK recommends that users keep their master branch updated, but to **NEVER** commit to it. Instead, it is highly preferred to do all your changes in a branch (branches are basically free in Git) and issue pull requests from your branches when you're developing.

There are instructions on how to keep your master updated here:

[**Best Practices: Your fork's master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_best_practices?id=your-fork39s-master-update-often-commit-never)

If you need any help with this just ask. **You do not need to close this pull request. Please continue to make all changes to your master branch until this PR is merged.**

codybender commented:
@noroadsleft this is great info, thank you so much for helping as I learn Git. I will be sure to branch my fork for any future changes

noroadsleft commented:
K, maybe it wasn't so difficult to fix. :)

noroadsleft commented:
@codybender,

If you're using Git through a command line interface, you'll want to make sure you have QMK's repo as an upstream repository (detailed on the doc I linked under the "Updating your master branch" section. Once you have that set up, run:

```
git checkout master
git fetch upstream
git reset --hard upstream/master
```

This will switch to your own master branch, fetch the references ("refs") from QMK's repository, and then change your local master branch to point to the same commit that QMK's master branch does. From here you can create new branches and commit to them, sync your repository; all the normal stuff for best practices. The normal sync method won't work for you because your master branch has commits that QMK's master doesn't, and Git will refuse to let you update because you'll lose those commits. (You don't actually lose them forever, but it would be difficult to find them again.)

There are some good resources on learning how to use Git in the docs.

[Learning Resources - QMK Firmware Documentation](https://docs.qmk.fm/#/newbs_learn_more_resources)

fauxpark commented on keyboards/keyhive/maypad/maypad.h:
```diff
@@ -0,0 +1,49 @@
+ */
+// The first section contains all of the arguments
+// The second converts the arguments into a two-dimensional array
+#define LAYOUT_numpad_5x4( \
```
Where did the `ortho_5x4` layout go? :( You should add it here and in info.json, as this board supports both.

codybender replied:
I removed it for simplicity, since I really had no idea what was going on. is it just a second define statement using `ortho_5x4`? And then the corresponding changes in the keymap

noroadsleft replied:
Basically, if someone builds a Maypad with, say, two 1u keys instead of the normal Numpad +, this gives them a macro that they can use in their keymap to access those positions on the PCB.

fauxpark replied:
Yep, but you don't need to do anything in the default keymap.c - it's so that people who aren't using the numpad layout can just use `LAYOUT_ortho_5x4(...)` in their keymaps. The keyboard should declare all possible layouts (within reason, some keyboards support *a lot* of configurations) so you can pick the one that matches how you've placed the switches.

fauxpark commented on keyboards/keyhive/maypad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,40 @@
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+	LAYOUT_numpad_5x4(
+		KC_NLCK, KC_PSLS, KC_PAST, KC_PMNS, \
```
The backslashes at the end of these lines are not necessary. You can keep them if you want as it isn't hurting anything, but IMO it looks neater without them.

codybender commented:
@noroadsleft is this something I should do now, or wait until after this PR is closed?

noroadsleft commented:
@codybender After this PR is merged; I failed to specify that. Good catch.

noroadsleft commented on keyboards/keyhive/maypad/maypad.h:
```diff
@@ -46,4 +46,16 @@
 }
 
-#define LAYOUT LAYOUT_numpad_5x4
+#define LAYOUT_ortho_5x4( \
```
I'd prefer if the identifiers were consistent between the two macros, which would make this:

```c
#define LAYOUT_ortho_5x4( \
    k00, k01, k02, k03, \
    k10, k11, k12, k13, \
    k20, k21, k22, k23, \
    k30, k31, k32, k33, \
    k40, k41, k42, k43  \
) { \
    { k00, k01, k02, k03 }, \
    { k10, k11, k12, k13 }, \
    { k20, k21, k22, k23 }, \
    { k30, k31, k32, k33 }, \
    { k40, k41, k42, k43 }  \
}
```

codybender replied:
Yeah, sorry I wasn't sure what numbers to fill in. I see now that it is k(row)(column), which makes WAY more sense.

drashna commented on keyboards/keyhive/maypad/rules.mk:
```diff
@@ -0,0 +1,82 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no      # Virtual DIP switch configuration(+1000)
```
Does this keyboard have a physical reset button?  And one that is relatively easy to access? 

if not, then you may want to enable bootmagic (either `full` or `lite`).

codybender replied:
It has an exposed Pro Micro, but no actual switch. Bootmagic looks like it will be a good option to have, though. I will get that added in!

drashna replied:
That's why I ask. :)   

Though, it may be simpler to go with "lite", here.  Since this is a macro pad (eg, not a fully keyboard), some of the advanced options for bootmagic won't be accessible.   So, to save on some space, `lite` just does the bootloader stuff.  And may be more suitable for this.

codybender replied:
I thought this may be the case. I will change it to `lite`! I appreciate the help

drashna commented on keyboards/keyhive/maypad/rules.mk:
```diff
@@ -0,0 +1,82 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = full      # Virtual DIP switch configuration(+1000)
```
This should probably be the "lite" version, since you can't even access some of the functionality, due to being a macro pad.  

So to save on a bit of space, it may be best to use the lite option, which just handles rebooting into the bootloader.

```suggestion
BOOTMAGIC_ENABLE = lite
```

drashna commented:
Looks good!

codybender commented:
Hi @drashna, I just realized after testing that `LAYOUT_numpad_5x4` that the pins were wrong for the 2u keys. I have fixed them and tested it, should I submit a separate PR to fix it? Not sure how this works

noroadsleft commented:
@codybender Yeah, submitting another PR with the fix would be the course of action, but as you haven't [reset your master branch yet](https://github.com/qmk/qmk_firmware/pull/6287#issuecomment-509826328), you might have issues with submitting another PR. The new PR may be in a merge conflict state, which could actually prevent us from merging it.

Perhaps go ahead and submit it and see what happens. In any case, I can walk you through getting the PR into a merge-ready state if it does end up with conflicts.

codybender commented:
@noroadsleft I can reset my master, and then put the changes into another branch with the fix if that would help!

noroadsleft commented:
@codybender If you can manage that it'd be easier and quicker on our end.

Once you get your master reset and your branch created and checked out, `git cherry-pick 122ceff4a1395b54ed7eddbc3b8644ff39acacde` should take the commit with your fix (currently on your master branch) and apply it to your new branch. From there you can post a PR.

(I remember you said you're still learning Git; there's a bunch of neat little commands and tricks for fixing these sorts of things.)

codybender commented:
@noroadsleft wow that cherry-pick command was very useful! it looks like I got it all ready. I really appreciate your help!
